### ISSUE TITLE
Define the OAuth scope for read access to subscription capabilities in profile

### ISSUE THREAD

**Issue by lmorchard** [mozilla/fxa#882]
Related to #695 

In PR #819, I used `profile:email` as the OAuth scope that grants access to subscription capabilities in a user's profile.

Do we want to instead define a specific scope just for subscription capabilities? (e.g. `https://identity.mozilla.com/apps/subscriptionCapabilities` or `profile:subscriptions` or suchlike?)

**Comment by clouserw** [comment]
We talked in triage and decided to use `profile:email` for now.  /cc @jvehent @gdestuynder @g-k in case they want us to change it in the future

**Comment by g-k** [comment]
> Do we want to instead define a specific scope just for subscription capabilities?

Yes, finer-grained permissions for the subscriptions data would be good. If we control the clients using `profile:email` and trust them with the subscriptions data that scope is OK for now.

**Comment by shane-tomlinson** [comment]
I'm re-opening this, overloading `profile:email` in this way conflates authorization for two distinct data types. It's easy for an RP that requests `profile:email` to automatically get subscription info, but I'm very hesitant to let this become *the* policy since it'd fall afoul of user expectations if an RP that only requested the `profile:email` scope were also to get information about the user's subscription status. We should follow the [principle of least surpise](http://principles-wiki.net/principles:principle_of_least_surprise) especially when it comes to user data.

I have to admit I'm also wondering whether using two seemingly unrelated scope names for reading (`profile:email`) and writing (`https://identity.mozilla.com/account/subscriptions`) is going to cause confusion. The writes should only be done by *us*, so maybe this isn't so bad.

Figuring out a coherent, fine grained scope name that can be used for either reading or writing might pay dividends in clarity and maintenance. I'm going to present two sides, using a [short name](https://github.com/mozilla/fxa/blob/master/packages/fxa-auth-server/fxa-oauth-server/docs/scopes.md#short-name-scope-values), and using a [URL name](https://github.com/mozilla/fxa/blob/master/packages/fxa-auth-server/fxa-oauth-server/docs/scopes.md#url-scopes), to see if one side makes more sense than the other.

### short name (profile:subscriptions)

Since this is data that is very much tied to a user's profile, adding a new suffix on the [`profile` short scope name](https://github.com/mozilla/fxa/blob/master/packages/fxa-auth-server/fxa-oauth-server/docs/scopes.md#profile-data) seems reasonable, @lmorchard's suggestion of `profile:subscriptions` looks great to me. Since short scope names are read-only by default, to write to the subscription data, an RP would need to request `profile:subscriptions:write` or `profile:write` scope. There are a couple of ways this gets a little weird though. First, `profile:write` usually writes data on the profile server, not the auth-server. Second, what happens if an RP is not allowed to actually write subscription data, IIUC the only services allowed to do that will be the subscription page service and the content server. Will requests for the `profile:write` grant cause an error when requesting the code/token, or would we wait until the access token is used to write subscription data? The latter seems most reasonable to me.

### URL name (https://identity.mozilla.com/account/subscriptions)

While subscription data is returned when fetching the user's profile information from the profile server, the write operations occur on the auth server rather than the profile server (maybe this implies *all the operations* should live on the profile server rather than the auth server, but hey...) Because write operations occur on a separate server, using a separate long URL scope makes sense for write operations. Forcing RPs to specify a long URL scope to get subscription info makes their /authorization request slightly longer, but there is a consistent scope name for read/write operations. Is a longer request for the RPs so bad? No, but I can imagine the confusion.

### Hybrid

Use a long name for write operations on the auth server, a short name for read operations on the profile server. That's what's happening now (yay) and after writing the above, I've re-convinced myself this is the best way forward. It makes it easy to prevent random RPs from requesting the long subscription management scope. However I'm still against overloading `profile:email`, a distinct scope name makes the most sense.

I'd like to propose:

* `profile:subscriptions` - Returns the subscription data relevant to the RP. 
  - If the RP has associated plans, return two bits of information: the plans the user is subscribed to that are relevant to the RP, and a boolean indicating whether the user has *any* active subscription.
  - If the RP has no associated plans, return a boolean indicating whether the user has *any* active subscriptions. This would be used for services like SUMO to determine whether to provide the user with premium support.
* `https://identity.mozilla.com/account/subscriptions` - For internal read/write operations, and should *only* be able to be requested by the content server and the subscription page.
  - Note this is different to both the previously mentioned `apps` and `acct`, this isn't an app, and cutting two characters out of account... meh...

**Comment by g-k** [comment]
Thanks for writing this up Shane! I really appreciate the clarity around the tradeoffs. 

+1 on the hybrid approach and not overloading `profile:email`

Also, if using a short and URL name ends up being confusing, we can reserve `profile:write` and `https://identity.mozilla.com/account/subscriptions#read` and have them return an error suggesting right name type. 

@gdestuynder any opinions or thoughts on this from your experience running our internal IdP?

cc @psiinon @ajvb additional opinions too

**Comment by gdestuynder** [comment]
In reply to @g-k 

**About Mozilla IAM/SSO:**
For reference this is how we implemented similar things:
https://github.com/mozilla-iam/cis/blob/master/docs/PersonAPI.md#what-are-the-access-controls-around-cis-person-api

We map the scope to a value in the database directly, simplified for illustration purposes here:
```
Document User:
{
 user_id: {
   value: user_id_here,
   meta_class: public
},
 subscription: {
  value: id_here,
  meta_class: workgroup_confidential
}
...
```
and the scopes to read both these would be `class:public class:workgroup_confidential`
or to write to subscription: `class:workgroup_confidential:write`

The reason we like this is that its very hard to make a mistake. On the flip side the scope is not super explicit about what the data is, only which type you're allowed to access, but that's a trade-off we actually like

**About proposals above**

I think what you've exposed above in the "Short name" proposal also makes sense to me. I would generally shy away from putting URL in scopes just because its super verbose and not very common so I would fear it brings some confusion.

**TLDR**  all proposals _except_ oversubscribing `profile:email` sound OK to me and offer different trade-offs.

**Comment by rfk** [comment]
+1 to your hybrid approach FWIW, I think it makes a lot of sense
